tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs

```diff
@@ -0,0 +1,609 @@
+                }
+            }
+
+            if (expectedEvent.TryGetValue("options", out var expectedOption))
+            {
+                var connectionPoolSettings = actualEvent.ConnectionPoolSettings();
+                if (expectedOption.IsInt32 && expectedOption == 42) // 42 - placeholder
+                {
+                    connectionPoolSettings.Should().NotBeNull();
+                }
+                else
+                {
+                    var expectedMaxPoolSize = expectedOption["maxPoolSize"].ToInt32();
+                    connectionPoolSettings.MaxConnections.Should().Be(expectedMaxPoolSize);
+                    var expectedMinPoolSize = expectedOption["minPoolSize"].ToInt32();
+                    connectionPoolSettings.MinConnections.Should().Be(expectedMinPoolSize);
+                    // maxIdleTimeMS : todo:
+                }
+            }
+
+            if (expectedEvent.TryGetValue("address", out var address))
+            {
+                var expectedAddress = address.ToInt32();
+                if (expectedAddress == 42)
+                {
+                    actualEvent.ServerId().EndPoint.Should().NotBeNull();
+                }
+                else
+                {
+                    // not expected code path
+                    throw new NotImplementedException();
+                }
+            }
+
+            if (expectedEvent.TryGetValue("reason", out var reason))
+            {
+                if (!(actualEvent is ConnectionClosedEvent)) // we don't support a reason for ConnectionClosedEvent
+                {
+                    actualEvent.CloseReason().Should().NotBeNull();
+                    actualEvent.CloseReason().ToString().ToLower().Should().Be(reason.ToString().ToLower());
+                }
+            }
+        }
+
+        private void AssertEvents(BsonDocument test, EventCapturer eventCapturer)
+        {
+            var actualEvents = GetFilteredEvents(eventCapturer, test);
+            var expectedEvents = GetExpectedEvents(test);
+            var minCount = Math.Min(actualEvents.Count, expectedEvents.Count);
+            for (var i = 0; i < minCount; i++)
+            {
+                var expectedEvent = expectedEvents[i];
+                JsonDrivenHelper.EnsureAllFieldsAreValid(expectedEvent, "type", "address", "connectionId", "options", "reason");
+                AssertEvent(actualEvents[i], expectedEvent);
+            }
+
+            if (actualEvents.Count < expectedEvents.Count)
+            {
+                throw new Exception($"Missing event: {expectedEvents[actualEvents.Count]}.");
+            }
+
+            if (actualEvents.Count > expectedEvents.Count)
+            {
+                throw new Exception($"Unexpected event of type: {actualEvents[expectedEvents.Count].GetType().Name}.");
+            }
+        }
+
+        private Task CreateTask(Action action)
```

Comment:
I've tried to use `Thread`s instead of `Task`s for this ticket. But without big results. I'm not sure how I can implement this logic: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.json&file-filters%5B%5D=.yml#diff-a638783da40724c2c11679e9321e1829R519

Also, the current implementation based on `tasks` looks good enough for me. I'm open to criticism/questions :)

Reply:
I'm fine with `Task`